toolchain/semantics/semantics_handle_index.cpp

```diff
@@ -2,20 +2,65 @@
 // Exceptions. See /LICENSE for license information.
 // SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
 
-#include <utility>
+#include <cstdint>
 
 #include "toolchain/semantics/semantics_context.h"
+#include "toolchain/semantics/semantics_node.h"
+#include "toolchain/semantics/semantics_node_kind.h"
 
 namespace Carbon {
 
-auto SemanticsHandleIndexExpressionStart(SemanticsContext& context,
-                                         ParseTree::Node parse_node) -> bool {
-  return context.TODO(parse_node, "HandleIndexExpressionStart");
+auto SemanticsHandleIndexExpressionStart(SemanticsContext& /*context*/,
+                                         ParseTree::Node /*parse_node*/)
+    -> bool {
+  // Leave the expression on the stack for IndexExpression.
+  return true;
 }
 
 auto SemanticsHandleIndexExpression(SemanticsContext& context,
                                     ParseTree::Node parse_node) -> bool {
-  return context.TODO(parse_node, "HandleIndexExpression");
+  auto index_node_id = context.node_stack().PopExpression();
+  auto index_node = context.semantics_ir().GetNode(index_node_id);
+  auto name_node_id = context.node_stack().PopExpression();
+  auto name_node = context.semantics_ir().GetNode(name_node_id);
+  auto name_type_id =
+      context.semantics_ir().GetTypeAllowBuiltinTypes(name_node.type_id());
+  auto name_type_node = context.semantics_ir().GetNode(name_type_id);
+
+  if (name_type_node.kind() == SemanticsNodeKind::TupleType &&
+      index_node.kind() == SemanticsNodeKind::IntegerLiteral) {
+    auto index_val = *(context.semantics_ir()
+                           .GetIntegerLiteral(index_node.GetAsIntegerLiteral())
+                           .getRawData());
+    auto type_block =
+        context.semantics_ir().GetTypeBlock(name_type_node.GetAsTupleType());
+
+    if (index_val >= static_cast<uint64_t>(type_block.size())) {
+      CARBON_DIAGNOSTIC(OutOfBoundsAccess, Error,
+                        "Index `{0}` is past the end of `{1}`.", uint64_t,
+                        std::string);
+      context.emitter().Emit(
+          parse_node, OutOfBoundsAccess, index_val,
```

Comment:
```suggestion
          parse_node, OutOfBoundsAccess, llvm::APSInt(index_val, /*isUnsigned=*/true),
```
We need to print as `APSInt` instead of as `APInt` so that we don't treat the high bit as a sign bit.